File: src/test/java/net/lumalyte/trivia/models/QuestionTest.java
package net.lumalyte.trivia.models;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

import java.util.Arrays;
import java.util.List;

public class QuestionTest {
    private Question multipleChoiceQuestion;
    private Question trueFalseQuestion;

    @BeforeEach
    void setUp() {
        // Set up a multiple choice question
        multipleChoiceQuestion = new Question(
            "What is the capital of France?",
            "Paris",
            Arrays.asList("London", "Berlin", "Madrid"),
            "Geography",
            "easy",
            "multiple"
        );

        // Set up a true/false question
        trueFalseQuestion = new Question(
            "The Earth is flat.",
            "False",
            List.of("True"),
            "Science",
            "easy",
            "boolean"
        );
    }

    @Test
    void testMultipleChoiceAnswerValidation() {
        // Test letter-based answers
        assertTrue(multipleChoiceQuestion.isCorrectAnswer(
            Character.toString((char)('a' + multipleChoiceQuestion.getShuffledAnswers().indexOf("Paris")))
        ));

        // Test case insensitivity
        assertTrue(multipleChoiceQuestion.isCorrectAnswer("PARIS"));
        assertTrue(multipleChoiceQuestion.isCorrectAnswer("paris"));
        assertTrue(multipleChoiceQuestion.isCorrectAnswer("Paris"));
    }

    @Test
    void testTrueFalseAnswerValidation() {
        // Test various ways to answer false
        assertTrue(trueFalseQuestion.isCorrectAnswer("f"));
        assertTrue(trueFalseQuestion.isCorrectAnswer("F"));
        assertTrue(trueFalseQuestion.isCorrectAnswer("false"));
        assertTrue(trueFalseQuestion.isCorrectAnswer("False"));
        assertTrue(trueFalseQuestion.isCorrectAnswer("FALSE"));

        // Test various ways to answer true (should be incorrect)
        assertFalse(trueFalseQuestion.isCorrectAnswer("t"));
        assertFalse(trueFalseQuestion.isCorrectAnswer("T"));
        assertFalse(trueFalseQuestion.isCorrectAnswer("true"));
        assertFalse(trueFalseQuestion.isCorrectAnswer("True"));
        assertFalse(trueFalseQuestion.isCorrectAnswer("TRUE"));
    }

    @Test
    void testFormattedAnswers() {
        // Test multiple choice formatting
        String mcAnswers = multipleChoiceQuestion.getFormattedAnswers();
        assertTrue(mcAnswers.contains("&eA)"));
        assertTrue(mcAnswers.contains("&eB)"));
        assertTrue(mcAnswers.contains("&eC)"));
        assertTrue(mcAnswers.contains("&eD)"));
        assertTrue(mcAnswers.contains("&f" + "Paris"));
        assertTrue(mcAnswers.contains("&f" + "London"));
        assertTrue(mcAnswers.contains("&f" + "Berlin"));
        assertTrue(mcAnswers.contains("&f" + "Madrid"));

        // Test true/false formatting
        String tfAnswers = trueFalseQuestion.getFormattedAnswers();
        assertTrue(tfAnswers.contains("&eA) &fTrue"));
        assertTrue(tfAnswers.contains("&eB) &fFalse"));
    }

    @Test
    void testGetCorrectAnswerLetter() {
        // Test true/false letter
        assertEquals("B", trueFalseQuestion.getCorrectAnswerLetter()); // False is always B

        // Test multiple choice letter
        String letter = multipleChoiceQuestion.getCorrectAnswerLetter();
        int index = letter.charAt(0) - 'A';
        assertEquals("Paris", multipleChoiceQuestion.getShuffledAnswers().get(index));
    }

    @Test
    void testGetters() {
        assertEquals("What is the capital of France?", multipleChoiceQuestion.getQuestion());
        assertEquals("Paris", multipleChoiceQuestion.getCorrectAnswer());
        assertEquals("Geography", multipleChoiceQuestion.getCategory());
        assertEquals("easy", multipleChoiceQuestion.getDifficulty());
        assertEquals("multiple", multipleChoiceQuestion.getType());
        assertEquals(3, multipleChoiceQuestion.getIncorrectAnswers().size());
        assertEquals(4, multipleChoiceQuestion.getShuffledAnswers().size());
    }
} 

File: src/test/kotlin/net/badgersmc/trivia/domain/QuestionTest.kt
package net.badgersmc.trivia.domain

import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class QuestionTest {
    private lateinit var multipleChoiceQuestion: Question
    private lateinit var trueFalseQuestion: Question

    @BeforeEach
    fun setUp() {
        multipleChoiceQuestion = Question(
            question = "What is the capital of France?",
            correctAnswer = "Paris",
            incorrectAnswers = listOf("London", "Berlin", "Madrid"),
            category = "Geography",
            difficulty = "easy",
            type = "multiple"
        )

        trueFalseQuestion = Question(
            question = "The Earth is flat.",
            correctAnswer = "False",
            incorrectAnswers = listOf("True"),
            category = "Science",
            difficulty = "easy",
            type = "boolean"
        )
    }

    @Test
    fun `letter answer is correct for multiple choice`() {
        val correctIndex = multipleChoiceQuestion.shuffledAnswers.indexOf("Paris")
        val letter = ('a' + correctIndex).toString()
        assertTrue(multipleChoiceQuestion.isCorrectAnswer(letter))
    }

    @Test
    fun `uppercase letter answer is correct`() {
        val correctIndex = multipleChoiceQuestion.shuffledAnswers.indexOf("Paris")
        val letter = ('A' + correctIndex).toString()
        assertTrue(multipleChoiceQuestion.isCorrectAnswer(letter))
    }

    @Test
    fun `wrong letter answer returns false`() {
        val correctIndex = multipleChoiceQuestion.shuffledAnswers.indexOf("Paris")
        val wrongIndex = if (correctIndex == 0) 1 else 0
        val letter = ('a' + wrongIndex).toString()
        assertFalse(multipleChoiceQuestion.isCorrectAnswer(letter))
    }

    @Test
    fun `case insensitive direct answer is correct`() {
        assertTrue(multipleChoiceQuestion.isCorrectAnswer("PARIS"))
        assertTrue(multipleChoiceQuestion.isCorrectAnswer("paris"))
        assertTrue(multipleChoiceQuestion.isCorrectAnswer("Paris"))
    }

    @Test
    fun `direct wrong answer returns false`() {
        assertFalse(multipleChoiceQuestion.isCorrectAnswer("London"))
    }

    @Test
    fun `true false T and F map correctly`() {
        assertTrue(trueFalseQuestion.isCorrectAnswer("f"))
        assertTrue(trueFalseQuestion.isCorrectAnswer("F"))
        assertTrue(trueFalseQuestion.isCorrectAnswer("false"))
        assertTrue(trueFalseQuestion.isCorrectAnswer("False"))
        assertTrue(trueFalseQuestion.isCorrectAnswer("FALSE"))

        assertFalse(trueFalseQuestion.isCorrectAnswer("t"))
        assertFalse(trueFalseQuestion.isCorrectAnswer("T"))
        assertFalse(trueFalseQuestion.isCorrectAnswer("true"))
    }

    @Test
    fun `shuffled answers contain correct answer`() {
        assertTrue(multipleChoiceQuestion.shuffledAnswers.contains("Paris"))
        assertEquals(4, multipleChoiceQuestion.shuffledAnswers.size)
    }

    @Test
    fun `formatted answers for multiple choice shows ABCD`() {
        val formatted = multipleChoiceQuestion.formattedAnswers
        assertTrue(formatted.contains("<yellow>A)"))
        assertTrue(formatted.contains("<yellow>B)"))
        assertTrue(formatted.contains("<yellow>C)"))
        assertTrue(formatted.contains("<yellow>D)"))
        assertTrue(formatted.contains("Paris"))
        assertTrue(formatted.contains("London"))
    }

    @Test
    fun `formatted answers for boolean shows True as A False as B`() {
        val formatted = trueFalseQuestion.formattedAnswers
        assertTrue(formatted.contains("<yellow>A) <white>True"))
        assertTrue(formatted.contains("<yellow>B) <white>False"))
    }

    @Test
    fun `correctAnswerLetter for boolean returns A or B`() {
        assertEquals("B", trueFalseQuestion.correctAnswerLetter)
    }

    @Test
    fun `correctAnswerLetter for multiple choice matches index`() {
        val letter = multipleChoiceQuestion.correctAnswerLetter
        val index = letter[0] - 'A'
        assertEquals("Paris", multipleChoiceQuestion.shuffledAnswers[index])
    }

    @Test
    fun `getters return constructor values`() {
        assertEquals("What is the capital of France?", multipleChoiceQuestion.question)
        assertEquals("Paris", multipleChoiceQuestion.correctAnswer)
        assertEquals("Geography", multipleChoiceQuestion.category)
        assertEquals("easy", multipleChoiceQuestion.difficulty)
        assertEquals("multiple", multipleChoiceQuestion.type)
        assertEquals(3, multipleChoiceQuestion.incorrectAnswers.size)
        assertEquals(4, multipleChoiceQuestion.shuffledAnswers.size)
    }
}
